src/Features/CSharp/Portable/CodeRefactorings/ConvertIfToSwitch/CSharpConvertToSwitchCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,293 @@
+            var sectionData = new List<(List<Pattern> patterns, StatementSyntax body)>();
+            var currentIfStatement = ifStatement;
+            while (true)
+            {
+                var operands = GetLogicalOrExpressionOperands(currentIfStatement.Condition).ToImmutableArray();
+                var patterns = new List<Pattern>(operands.Length);
+                foreach (var operand in operands)
+                {
+                    var pattern = GetPattern(operand);
+                    if (pattern == null)
+                    {
+                        return;
+                    }
+
+                    patterns.Insert(0, pattern);
+                }
+
+                sectionData.Add((patterns, currentIfStatement.Statement));
+                elseBody = currentIfStatement.Else?.Statement;
+                currentIfStatement = elseBody as IfStatementSyntax;
+                if (currentIfStatement == null)
+                {
+                    break;
+                }
+            }
+
+            context.RegisterRefactoring(new MyCodeAction(c =>
+            {
+                IEnumerable<SyntaxNode> GetSwitchSectionBody(StatementSyntax node)
+                {
+                    if (!(node is BlockSyntax block && block.Statements.Count == 0))
+                    {
+                        yield return node;
+                    }
+
+                    if (semanticModel.AnalyzeControlFlow(node).EndPointIsReachable)
+                    {
+                        yield return SyntaxFactory.BreakStatement();
+                    }
+                }
+
+                CasePatternSwitchLabelSyntax GetSwitchLabel(Pattern pattern)
+                {
+                    CasePatternSwitchLabelSyntax CasePatternSwitchLabel(TypeSyntax type, VariableDesignationSyntax designation)
+                        => SyntaxFactory.CasePatternSwitchLabel(
+                            SyntaxFactory.DeclarationPattern(type, designation),
+                            SyntaxFactory.Token(SyntaxKind.ColonToken));
+
+                    switch (pattern)
+                    {
+                        case Pattern.ByValue p:
+                            return SyntaxFactory.CasePatternSwitchLabel(
+                                SyntaxFactory.ConstantPattern(p.Expression),
+                                SyntaxFactory.Token(SyntaxKind.ColonToken));
+
+                        case Pattern.ByType p:
+                            return CasePatternSwitchLabel(p.Type, SyntaxFactory.SingleVariableDesignation(p.Identifier));
+
+                        case Pattern.Discarded p:
+                            return CasePatternSwitchLabel(p.Type, SyntaxFactory.DiscardDesignation());
+
+                        case Pattern.Guarded p:
+                            return GetSwitchLabel(p.Pattern).WithWhenClause(SyntaxFactory.WhenClause(p.Guard));
+
+                        default:
+                            Debug.Assert(false);
```

Comment:
I couldn't find ExceptionUtilities.Unreachable in IDE projects, so there is that. Will change.